p8e-api/src/main/kotlin/io/provenance/engine/stream/EventStream.kt

```diff
@@ -0,0 +1,288 @@
+                } else {
+                    it.usePlaintext()
+                }
+            }
+            .maxInboundMessageSize(20 * 1024 * 1024) // ~ 20 MB
+            .idleTimeout(5, TimeUnit.MINUTES)
+            .keepAliveTime(60, TimeUnit.SECONDS) // ~ 12 pbc block cuts
+            .keepAliveTimeout(20, TimeUnit.SECONDS)
+            .build()
+
+        ABCIApplicationGrpc.newBlockingStub(channel)
+    }
+
+    fun getStream(eventTypes: List<String>, startHeight: Long, observer: EventStreamResponseObserver<EventBatch>) = EventStream(URI(eventStreamProperties.websocketUri), eventTypes, startHeight, observer, transactionQueryService)
+
+    class EventStream(
+        val node: URI,
+        val eventTypes: List<String>,
+        val startHeight: Long,
+        private val observer: EventStreamResponseObserver<EventBatch>,
+        private val transactionQueryService: TransactionQueryService
+    ) {
+        companion object {
+            private val HISTORY_BATCH_SIZE = 10
+            private val executor = ThreadPoolFactory.newFixedThreadPool(HISTORY_BATCH_SIZE, "event-stream-%d")
+        }
+
+        private val log = logger()
+
+        private val lifecycle = LifecycleRegistry(0L)
+        private val scarletInstance = Scarlet.Builder()
+            .lifecycle(lifecycle)
+            .webSocketFactory(OkHttpClient.Builder()
+                .readTimeout(Duration.ofSeconds(60)) // ~ 12 pbc block cuts
+                .build()
+                .newWebSocketFactory("${node.scheme}://${node.host}:${node.port}/websocket"))
+            .addMessageAdapterFactory(MoshiMessageAdapter.Factory())
+            .addStreamAdapterFactory(RxJava2StreamAdapterFactory())
+            .build()
+        private val eventStreamService = scarletInstance.create<EventStreamService>()
+
+        private var subscription: Disposable? = null
+
+        private var lastBlockSeen = AtomicLong(-1)
+        private val eventMonitor = thread(start = false) {
+            var lastBlockMonitored = lastBlockSeen.get()
+            while (true) {
+                Thread.sleep(30000)
+                val lastBlockSeen = lastBlockSeen.get()
+                log.debug("Checking for event stream liveliness [lastBlockSeen: $lastBlockSeen vs. lastBlockMonitored: $lastBlockMonitored]")
+                if (lastBlockSeen <= lastBlockMonitored) {
+                    handleError(EventStreamStaleException("EventStream has not received a block in 30 seconds [lastBlockSeen: $lastBlockSeen vs. lastBlockMonitored: $lastBlockMonitored]"))
+                    break
+                }
+                lastBlockMonitored = lastBlockSeen
+            }
+            log.info("Exiting event monitor thread")
+        }
+
+        fun streamEvents() {
+            // todo: need to limit how many times this function is called??? Used to limit based on consumer id... probably need to use redis to do this... and ensure cleaned up when shutting down if do
```

Comment:
I think we probably want to do this (prevent an EventStream running simultaneously on multiple instances of p8e with the same consumer id so we don't double-process events when deploying, etc)...

the old event-stream could easily do this as it was just one node and kept track of the connected consumers, but we would either need to use Redis (which I know we have discussed getting rid of) or maybe have a heartbeat that touches some record in a db table signaling that it is still alive... thoughts on an approach here?

Reply:
I think adding a new component (Redis) fo a small piece would be overkill? I'd prefer the heartbeat idea with a flag somewhere. 

Does the LifecycleRegistery provide consumer information that we can utilize?

Reply:
I think the LifecycleRegistry is really just a state machine that allows to control when the WebSocket closes/opens, so I don't think it could facilitate communication between instances

Reply:
We can start keeping track of these as `TODO: concurrency-` or something that way we can find and solve for them at once.

Reply:
added a concurrency note.

I looked into this a bit more and realized we were always using an empty string as our 'consumer id', so the event stream was actually considering us an anonymous consumer and not enforcing the single connection anyways, so leaving this as-is will be no different than it currently is